### PR TITLE
Lauren & Isaac | Carets

### PR DESCRIPTION
**Pull request by hisaacdelr**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  Scoring methods are called in both Player and Game classes, but in both cases, the callers don't need to know about any particular instance of Scoring. Player and Game are providing arguments (a word or array of words) and receiving the result of Scoring's methods, without having to know or keep track of any instance of Scoring. It helps with the Single-Responsibility principle.  |
|   Describe an example of a test you wrote for a _nominal case_. |  In our Scoring class, our first test "`Scrabble::Scoring.score(“hello”).must_equal 8`" because it does the bare minimum of the self.score method, which is, given a word, it should return it’s score.	|
|   Describe an example of a test you wrote for an _edge case_. |  For our Scoring class, we wrote a test for the `highest_score_from` method that requires an error to be raised in case any of the array elements are not an alphabetic string. This is an edge case because the method would have to be misused (a non-valid string is played by a player) in order for that code to run. 	|
|   What was your final test coverage percentage?	|   100% according to the RSpec Coverage report	|
|   Is there a specific piece of code you'd like feedback on?	|   Lauren: I would like feedback on how to make the `Scoring.highest_score_from` method cleaner and more readable. Isaac: I would like feedback on the logic in the Board method `can_place_word_on_board?` on how it can be less redundant and more readable.	|

### CODE REVIEW

**Comment by tildeee** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  x  |
|   Both Teammates contributed to the codebase	|  x  |
|   Regular Commits with meaningful commit messages  |  x  |
|  **Scoring class**	|   	|
|  `score` method	|  x  |
|   Uses appropriate data structure to store the letter score	|  x  |
|   Appropriately handles edge cases	|  x  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  x  |
|  `highest_scoring_word` method	|  x  |
|   Appropriately handles edge cases	|  x  |
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|  x  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  x  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  x  |
|  Returns `true` or `false` from `won?` method |  x  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |   relied on the comprehensive tests in score (this is good!)  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  x  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  x  |
| `tiles_remaining` method returns a count	|  x  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|   	|

File: specs/scoring_spec.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    it "returns the first word in the array when all words have the same points and length" do
+      words = ["dog", "dug", "dig", "God"]
+      Scrabble::Scoring.highest_score_from(words).must_equal "dog"
+    end
+
+    it "raises an error if one of the words in the array is not composed of only letters" do
+      words1 = ["dog", "*@?;!^", ["b", "y", "e"], ""]
+      words2 = ["*@?;!^", "apple", "!@%"]
+      words3 = ["banana", "pine", nil]
```
Comment (tildeee) [inline specs/scoring_spec.rb:57]: in general, your tests are comprehensive and great! I love them, these tests are thorough and easy to read.

File: lib/player.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      @tiles = []
+    end
+
+    def play(word)
+      if won?
+        return false
+      else
+        plays << word
+        Scoring.score(word)
+      end
```
Comment (tildeee) [inline lib/player.rb:18]: this method forgets to take out the letters from the player's `@tiles` (not necessary for any requirements)

**Comment by tildeee** [comment]
Great code, works really well and looks really good. I'm really pleased with your style of test writing, too-- I thought they were very thorough and covered a lot of cases.

Excellent job with the optionals! The board looks great, and I was really happy to run your `scrabble-with-board.rb` ! That was a lot of fun and it was very impressive. Great work, you two!



Re: your comprehension questions asking for feedback:
For `Scoring#self.highest_score_from`, my solution has fifteen lines while yours has nine, so I actually prefer your solution! Hahaha.
For `can_place_word_on_board?`: I actually think this well done and works with the 2D array really well. I thought the helper methods you created for it, particularly `convert_index_coordinate_to_tile`, were really clever.
There's no real way to get around this nested code if you're using a 2D array (which is the direction I thought people who implemented `Board` would go). But the code style in here definitely made it easy to understand with its well named helper methods and post-fix conditionals.
Unless you two underwent a major refactoring of the board and didn't make it a 2D array, but instead created a bunch of `Tile` objects that had references to other `Tile`s that were next to them in each direction, I can't think of many other ways to execute this. (Even if you made `Tile` objects that pointed to `left` `right` `up` `down` I'm not sure that'd even be a worthwhile solution)

If you'd like for me to sit down with you two to look over this code more in depth let me know.

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
+require 'terminal-table'
+require 'colorize'
+
+module Scrabble
+  class Board
+
+    BOARD_SIDE_SIZE = 15
```
Comment (tildeee) [inline lib/board.rb:7]: I'm really happy that it was an obvious choice to make this a constant, instead of hardcoding a size
